src/goldilock.cpp

```diff
@@ -1,54 +1,209 @@
 // Copyright 2024 Yannic Staudt, tipi technologies Ltd and the goldilock contributors
 // SPDX-License-Identifier: GPL-2.0-only OR Proprietary
 
+#include <chrono>
 #include <iostream>
-#include <boost/filesystem.hpp>
+#include <map>
+#include <optional>
 #include <string>
-#include <vector>
 #include <thread>
-#include <chrono>
+#include <vector>
+
+#include <boost/asio.hpp>
+#include <boost/predef.h>
+#include <boost/filesystem.hpp>
+#include <boost/process.hpp>
+#include <boost/process/handles.hpp>
+#include <boost/scope_exit.hpp>
+#if BOOST_OS_WINDOWS
+#include <boost/winapi/process.hpp>
+#endif
 
 #include <cxxopts.hpp>
+
 #include <goldilock/file.hpp>
 #include <goldilock/fstream.hpp>
 #include <goldilock/process_info.hpp>
 #include <goldilock/string.hpp>
+#include <goldilock/goldilock_spot.hpp>
+#include <goldilock/version.hpp> // generated by build script - located in binary dir
+
 
 namespace tipi::goldilock
 {  
-  std::string getenv_or_default(const char* var_name, const std::string& default_value = "") {
-    auto val = std::getenv(var_name);
+  namespace fs = boost::filesystem;
+  namespace bp = boost::process;
+  using namespace std::string_literals;
+  using namespace std::chrono_literals;
 
-    if(val != nullptr) {
-      return std::string(val);
-    }
+  inline std::ostream nowhere_sink(0);
 
-    return default_value;
+  #if BOOST_OS_WINDOWS
+  // boost process handler to tell windows to create a new console when setting up
+  // the new process which is the best equivalent of running a detached process
+  struct new_window_handler : ::boost::process::detail::handler_base
+  {
+    template <class Executor>
+    void on_setup(Executor &e) const
+    {
+      e.creation_flags = boost::winapi::CREATE_NEW_CONSOLE_;
+    }      
+  };
+  #endif
+
+  //!\brief run a command in a bp::child() using the system shell (cmd/bash)
+  template<class... Args>
+  inline auto shell_run(Args&&...args)
```

Comment:
this function and `new_window_handler` should go into a separate include file as it could be useful to have them in a reusable way